testsuite/python/p3m_madelung.py

```diff
@@ -0,0 +1,294 @@
+            pos = np.array([j, k, 0])
+            self.system.part.add(pos=spacing * pos, dip=[0., 0., 1.])
+
+        dp3m_params = dict(prefactor=1., accuracy=1e-9, mesh=48, r_cut=33.6,
+                           cao=7, tuning=False)
+        mdlc_params = {'maxPWerror': 1e-9, 'gap_size': 16.}
+
+        def check():
+            # minimal energy configuration
+            self.system.part.all().dip = dipm * np.array(dipoles_parallel)
+            mc = -2.54944
+            energy_per_dip = self.get_normalized_obs_per_dipole(dipm, spacing)
+            np.testing.assert_allclose(energy_per_dip, mc, atol=0., rtol=4e-6)
+
+            # maximal energy configuration
+            self.system.part.all().dip = dipm * np.array(dipoles_antiparallel)
+            mc = +3.01716
+            energy_per_dip = self.get_normalized_obs_per_dipole(dipm, spacing)
+            np.testing.assert_allclose(energy_per_dip, mc, atol=0., rtol=4e-6)
+
+        dp3m = espressomd.magnetostatics.DipolarP3M(**dp3m_params)
+        self.system.actors.add(dp3m)
+        check()
+        self.system.actors.remove(dp3m)
+        mdlc = espressomd.magnetostatics.DLC(actor=dp3m, **mdlc_params)
+        self.system.actors.add(mdlc)
+        check()
+
+    @utx.skipIfMissingFeatures(["DP3M"])
+    def test_infinite_magnetic_cube(self):
+        n_part = 8
+        spacing = 2.4
+        dipm = 1.05
+        self.system.box_l = 3 * [n_part * spacing]
+
+        dipoles_parallel = []
+        dipoles_antiparallel = []
+        for i, j, k in itertools.product(range(n_part), repeat=3):
+            if i % 2 == 0:
+                theta = np.pi + (-1)**j * np.pi / 4.
+            else:
+                theta = (-1)**(j + 1) * np.pi / 4.
+            dipoles_antiparallel.append([np.cos(theta), np.sin(theta), 0.])
+            dipoles_parallel.append([0., 0., (-1)**(i + j)])
+            pos = np.array([i, j, k])
+            self.system.part.add(pos=spacing * pos, dip=[0., 0., 1.])
+
+        dp3m_params = dict(
+            prefactor=1., accuracy=5e-8, mesh=64, r_cut=8.2, cao=7, tuning=False)
+
+        def check():
+            # minimal energy configuration
+            self.system.part.all().dip = dipm * np.array(dipoles_parallel)
+            mc = -2.67679
+            energy_per_dip = self.get_normalized_obs_per_dipole(dipm, spacing)
+            np.testing.assert_allclose(energy_per_dip, mc, atol=0., rtol=1e-6)
+
+            # maximal energy configuration
+            self.system.part.all().dip = dipm * np.array(dipoles_antiparallel)
+            mc = +4.84473  # imprecise
+            energy_per_dip = self.get_normalized_obs_per_dipole(dipm, spacing)
+            np.testing.assert_allclose(energy_per_dip, mc, atol=0., rtol=3e-4)
+
+        dp3m = espressomd.magnetostatics.DipolarP3M(**dp3m_params)
+        self.system.actors.add(dp3m)
+        check()
+
+    @utx.skipIfMissingFeatures(["P3M"])
+    def test_infinite_ionic_wire(self):
+        n_pairs = 64
```

Comment:
what is the reason for using that many pairs of ions? The result should be independent on the number of ion pairs...

Reply:
The precision of the energy and stress tensor is between 2 and 3 orders of magnitude larger than the P3M accuracy, therefore to test with a tolerance of 1e-6, the P3M accuracy needs to be in the range 1e-8 to 1e-9. But in 1D it's really hard to find a r_cut and cao that allows for such a precision. I'm not really sure why adding more particles, because the real and k-space error magnitudes should actually increase with the square root of the number of particles. Yet in the 3D case you can get away with 8 pairs. Maybe the numerical error gets down when you do the reduction to get the energy or stress, since you divide by the number of particles, i.e. the error gets overall divided by the square root of the number of particles?